File: swift/Sources/CircleAI/Companion.swift
// Companion.swift
//
// InterfaceKind, CompanionContext, CompanionTurn, CompanionProactiveEvent,
// and the ICompanionSession protocol.
// The Companion is the HER + JARVIS persona — available on every surface,
// with memory and identity that travels with the person.

import Foundation

// MARK: - InterfaceKind

/// The surface on which the Companion session is running.
/// Determines sensory capabilities, available UI affordances, and
/// how the Companion adapts its communication style.
public enum InterfaceKind: String, Sendable, CaseIterable {
    /// Mobile phone or tablet (MAUI).
    case mobile
    /// Smartwatch or fitness band with a small display.
    case wearable
    /// Desktop or laptop computer (MAUI or WPF).
    case desktop
    /// Browser-based experience (Blazor).
    case web
    /// Embedded IoT device — voice in, voice out, minimal compute.
    case iot
    /// Always-on ambient surface — smart speaker, room display, car.
    case ambient
    /// Programmatic / background / testing context (no UI).
    case headless
}

// MARK: - CompanionContext

/// Snapshot of all context injected into the Companion's system prompt.
/// Rebuilt at the start of each session and refreshed on request.
public struct CompanionContext: Sendable {
    public var identityId: String
    public var displayName: String
    public var preferredLanguage: String?
    public var interface: InterfaceKind
    public var personaHints: String
    public var affectSummary: String
    public var recentMemorySnippets: [String]
    public var activeGoals: [String]
    public var contextBuiltAt: Date

    public init(
        identityId: String,
        displayName: String,
        preferredLanguage: String? = nil,
        interface: InterfaceKind,
        personaHints: String,
        affectSummary: String,
        recentMemorySnippets: [String],
        activeGoals: [String],
        contextBuiltAt: Date = Date()
    ) {
        self.identityId = identityId
        self.displayName = displayName
        self.preferredLanguage = preferredLanguage
        self.interface = interface
        self.personaHints = personaHints
        self.affectSummary = affectSummary
        self.recentMemorySnippets = recentMemorySnippets
        self.activeGoals = activeGoals
        self.contextBuiltAt = contextBuiltAt
    }
}

// MARK: - CompanionTurn

/// A single turn in the Companion conversation log, held in memory for the
/// duration of the session.
public struct CompanionTurn: Sendable {
    /// "user" or "assistant"
    public var role: String
    public var content: String
    public var timestamp: Date

    public init(role: String, content: String, timestamp: Date = Date()) {
        self.role = role
        self.content = content
        self.timestamp = timestamp
    }
}

// MARK: - CompanionProactiveEvent

/// Metadata emitted when the Companion proactively initiates contact.
public struct CompanionProactiveEvent: Sendable {
    public var sessionId: String
    public var identityId: String
    public var interface: InterfaceKind
    public var message: String
    public var triggerName: String
    public var generatedAt: Date

    public init(
        sessionId: String,
        identityId: String,
        interface: InterfaceKind,
        message: String,
        triggerName: String,
        generatedAt: Date = Date()
    ) {
        self.sessionId = sessionId
        self.identityId = identityId
        self.interface = interface
        self.message = message
        self.triggerName = triggerName
        self.generatedAt = generatedAt
    }
}

// MARK: - FaceAffectMapper

/// Applies facial expression observations to an AffectState.
/// Called once per detected frame; the AffectState is mutated in-place.
///
/// Exact deltas (must match fixtures/facex_biometric_vectors.json):
///   happy:     engagement += 0.03, energy     += 0.02
///   surprised: curiosity  += 0.04
///   confused:  uncertainty += 0.05
///   stressed:  uncertainty += 0.08, energy    -= 0.05
///   angry:     engagement  -= 0.04, rapport   -= 0.02
///   all others: no change
///
/// Frames with confidenceScore < 0.5 are discarded entirely.
public enum FaceAffectMapper {

    /// Applies the expression delta from matrix to affect.
    /// No-op when matrix.confidenceScore < 0.5.
    public static func apply(_ matrix: FacialMetricMatrix, to affect: AffectState) {
        guard matrix.confidenceScore >= 0.5 else { return }
        switch matrix.expression {
        case .happy:
            affect.engagement = min(1.0, affect.engagement + 0.03)
            affect.energy     = min(1.0, affect.energy     + 0.02)
        case .surprised:
            affect.curiosity  = min(1.0, affect.curiosity  + 0.04)
        case .confused:
            affect.uncertainty = min(1.0, affect.uncertainty + 0.05)
        case .stressed:
            affect.uncertainty = min(1.0, affect.uncertainty + 0.08)
            affect.energy      = max(0.0, affect.energy      - 0.05)
        case .angry:
            affect.engagement = max(0.0, affect.engagement - 0.04)
            affect.rapport    = max(0.0, affect.rapport    - 0.02)
        default:
            return
        }
        affect.lastUpdatedAt = Date()
    }
}

// MARK: - FaceCompanionBridge

/// Bridges the face analysis pipeline to the Companion session.
/// Applies affect updates and emits proactive events when confusion is detected.
public enum FaceCompanionBridge {

    /// Uncertainty threshold above which a confusion proactive event is emitted.
    public static let confusionThreshold: Float = 0.70

    /// Observes a facial metric matrix, applies it to the affect state,
    /// and returns a proactive event if the confusion threshold is crossed
    /// while the expression is .confused or .stressed.
    ///
    /// - Returns: A `CompanionProactiveEvent` when the confusion threshold is
    ///   met, otherwise nil.
    public static func observe(
        _ matrix: FacialMetricMatrix,
        affect: AffectState,
        sessionId: String,
        identityId: String,
        surface: InterfaceKind
    ) -> CompanionProactiveEvent? {
        FaceAffectMapper.apply(matrix, to: affect)
        let crossed = affect.uncertainty >= confusionThreshold &&
            (matrix.expression == .confused || matrix.expression == .stressed)
        guard crossed else { return nil }
        return CompanionProactiveEvent(
            sessionId: sessionId,
            identityId: identityId,
            interface: surface,
            message: "I notice you might be finding this a bit tricky. Would you like me to slow down or explain it differently?",
            triggerName: "face.confusion_detected",
            generatedAt: Date()
        )
    }
}

// MARK: - ICompanionSession

/// A Companion conversation session. Combines identity awareness, cross-device
/// memory, language adaptation, affect sensing, and proactive reasoning into a
/// single coherent interface.
public protocol ICompanionSession: AnyObject {

    // ── Identity ──────────────────────────────────────────────────────────

    /// Stable unique identifier for this session.
    var sessionId: String { get }

    /// The authenticated identity driving this session.
    var identityId: String { get }

    /// The surface on which this session is running.
    var interface: InterfaceKind { get }

    // ── Core conversation ─────────────────────────────────────────────────

    /// Send a message to the Companion and receive a complete reply.
    func send(_ message: String) async throws -> String

    /// Stream the Companion's reply token-by-token for low-latency rendering.
    func stream(_ message: String) -> AsyncStream<String>

    /// Agentic mode: sends the instruction, detects tool calls, executes them,
    /// and re-prompts until the model produces a plain-text answer.
    func agent(_ instruction: String) async throws -> String

    // ── Context ───────────────────────────────────────────────────────────

    /// Returns the most recent CompanionContext snapshot.
    func getContext() -> CompanionContext

    /// Refreshes the context from backing stores.
    func refreshContext() async throws

    // ── History ───────────────────────────────────────────────────────────

    /// The in-session conversation history (this session only, not persisted).
    var history: [CompanionTurn] { get }

    // ── Feedback ──────────────────────────────────────────────────────────

    /// Signal satisfaction with the last reply.
    func signalFeedback(positive: Bool, note: String?) async throws

    // ── Proactive ─────────────────────────────────────────────────────────

    /// Stream of proactive events the Companion initiates without being prompted.
    var proactiveEvents: AsyncStream<CompanionProactiveEvent> { get }
}

File: android/circle-ai/src/main/kotlin/com/bhengubv/circleai/android/companion/Companion.kt
// Companion.kt
//
// Android/Kotlin port of Circle.AI.Companion portable layer.
//
// Covers:
//   InterfaceKind              — enum of surfaces (Mobile, Wearable, Desktop, Web, IoT, Ambient, Headless)
//   CompanionContext           — snapshot of all context injected into the system prompt
//   CompanionTurn              — a single turn in the session conversation log
//   CompanionProactiveEvent    — metadata emitted on proactive companion contact
//   FaceAffectMapper           — applies facial expression deltas to AffectState
//   FaceCompanionBridge        — observes face metrics and emits proactive events
//   ICompanionSession          — primary contract for a Circle AI Companion session

package com.bhengubv.circleai.android.companion

import com.bhengubv.circleai.android.memory.AffectState
import com.bhengubv.circleai.android.tools.FaceExpressionClassification
import com.bhengubv.circleai.android.tools.FacialMetricMatrix
import kotlinx.coroutines.flow.Flow
import java.time.Instant

// ---------------------------------------------------------------------------
// InterfaceKind
// ---------------------------------------------------------------------------

/**
 * The surface on which the Companion session is running.
 */
enum class InterfaceKind {
    /** Mobile phone or tablet (MAUI). */
    Mobile,
    /** Smartwatch or fitness band with a small display. */
    Wearable,
    /** Desktop or laptop computer (MAUI or WPF). */
    Desktop,
    /** Browser-based experience (Blazor). */
    Web,
    /** Embedded IoT device — voice in, voice out, minimal compute. */
    IoT,
    /** Always-on ambient surface — smart speaker, room display, car. */
    Ambient,
    /** Programmatic / background / testing context (no UI). */
    Headless,
}

// ---------------------------------------------------------------------------
// CompanionContext
// ---------------------------------------------------------------------------

/**
 * Snapshot of all context injected into the Companion's system prompt.
 */
data class CompanionContext(
    val identityId: String,
    val displayName: String,
    val preferredLanguage: String?,
    val interfaceKind: InterfaceKind,
    val personaHints: String,
    val affectSummary: String,
    val recentMemorySnippets: List<String>,
    val activeGoals: List<String>,
    val contextBuiltAt: Instant
)

// ---------------------------------------------------------------------------
// CompanionTurn
// ---------------------------------------------------------------------------

/**
 * A single turn in the Companion conversation log.
 * [role] is "user" or "assistant".
 */
data class CompanionTurn(
    val role: String,
    val content: String,
    val timestamp: Instant
)

// ---------------------------------------------------------------------------
// CompanionProactiveEvent
// ---------------------------------------------------------------------------

/**
 * Metadata emitted when the Companion proactively initiates contact.
 */
data class CompanionProactiveEvent(
    val sessionId: String,
    val identityId: String,
    val interfaceKind: InterfaceKind,
    val message: String,
    val triggerName: String,
    val generatedAt: Instant
)

// ---------------------------------------------------------------------------
// FaceAffectMapper
// ---------------------------------------------------------------------------

/**
 * Applies facial expression deltas to an [AffectState].
 *
 * CRITICAL: The delta constants below must match the reference implementation
 * exactly — they are cross-validated against fixtures/facex_biometric_vectors.json.
 * Confidence scores below [MIN_CONFIDENCE] are discarded without mutation.
 */
object FaceAffectMapper {

    private const val MIN_CONFIDENCE = 0.5f

    /**
     * Applies expression-driven affect deltas to [affect].
     * No-op if [matrix.confidenceScore] < [MIN_CONFIDENCE].
     *
     * Deltas (all values coerced to [0f, 1f] after application):
     *   Happy     → engagement +0.03, energy +0.02
     *   Surprised → curiosity  +0.04
     *   Confused  → uncertainty +0.05
     *   Stressed  → uncertainty +0.08, energy -0.05
     *   Angry     → engagement -0.04, rapport -0.02
     *   All other expressions (Neutral, Sad, Unknown) → no change
     */
    fun apply(matrix: FacialMetricMatrix, affect: AffectState) {
        if (matrix.confidenceScore < MIN_CONFIDENCE) return

        when (matrix.expression) {
            FaceExpressionClassification.Happy -> {
                affect.engagement = (affect.engagement + 0.03f).coerceIn(0f, 1f)
                affect.energy     = (affect.energy     + 0.02f).coerceIn(0f, 1f)
            }
            FaceExpressionClassification.Surprised ->
                affect.curiosity = (affect.curiosity + 0.04f).coerceIn(0f, 1f)
            FaceExpressionClassification.Confused ->
                affect.uncertainty = (affect.uncertainty + 0.05f).coerceIn(0f, 1f)
            FaceExpressionClassification.Stressed -> {
                affect.uncertainty = (affect.uncertainty + 0.08f).coerceIn(0f, 1f)
                affect.energy      = (affect.energy      - 0.05f).coerceIn(0f, 1f)
            }
            FaceExpressionClassification.Angry -> {
                affect.engagement = (affect.engagement - 0.04f).coerceIn(0f, 1f)
                affect.rapport    = (affect.rapport    - 0.02f).coerceIn(0f, 1f)
            }
            else -> return
        }

        affect.lastUpdatedAt = Instant.now()
    }
}

// ---------------------------------------------------------------------------
// FaceCompanionBridge
// ---------------------------------------------------------------------------

/**
 * Observes a stream of [FacialMetricMatrix] frames, applies them to
 * [AffectState] via [FaceAffectMapper], and emits a [CompanionProactiveEvent]
 * when the user appears confused or stressed beyond [CONFUSION_THRESHOLD].
 */
object FaceCompanionBridge {

    /**
     * Uncertainty threshold above which a confusion/stress event is emitted.
     * Applied after the face-driven affect delta has been applied.
     */
    const val CONFUSION_THRESHOLD = 0.70f

    /**
     * Observes one [FacialMetricMatrix] frame. Applies affect deltas then
     * returns a [CompanionProactiveEvent] if the confusion threshold has been
     * crossed, or null otherwise.
     *
     * @param matrix     The facial metric data for this frame.
     * @param affect     The mutable affect state to update in-place.
     * @param sessionId  Companion session identifier for the event payload.
     * @param identityId Identity that owns this session.
     * @param surface    The interface kind of the running session.
     */
    fun observe(
        matrix: FacialMetricMatrix,
        affect: AffectState,
        sessionId: String,
        identityId: String,
        surface: InterfaceKind
    ): CompanionProactiveEvent? {
        FaceAffectMapper.apply(matrix, affect)

        val isConfusionExpression = matrix.expression == FaceExpressionClassification.Confused ||
            matrix.expression == FaceExpressionClassification.Stressed

        val crossed = affect.uncertainty >= CONFUSION_THRESHOLD && isConfusionExpression

        if (!crossed) return null

        return CompanionProactiveEvent(
            sessionId    = sessionId,
            identityId   = identityId,
            interfaceKind = surface,
            message      = "I notice you might be finding this a bit tricky. " +
                           "Would you like me to slow down or explain it differently?",
            triggerName  = "face.confusion_detected",
            generatedAt  = Instant.now()
        )
    }
}

// ---------------------------------------------------------------------------
// ICompanionSession
// ---------------------------------------------------------------------------

/**
 * A Companion conversation session. Combines identity awareness, cross-device
 * memory, language adaptation, affect sensing, and proactive reasoning into a
 * single coherent interface.
 */
interface ICompanionSession : AutoCloseable {

    // ── Identity ─────────────────────────────────────────────────────────────

    /** Stable unique identifier for this session. */
    val sessionId: String

    /** The authenticated identity driving this session. */
    val identityId: String

    /** The surface on which this session is running. */
    val interfaceKind: InterfaceKind

    // ── Core conversation ─────────────────────────────────────────────────────

    /**
     * Send a message to the Companion and receive a complete reply.
     */
    suspend fun sendAsync(message: String): String

    /**
     * Stream the Companion's reply token-by-token for low-latency rendering.
     */
    fun streamAsync(message: String): Flow<String>

    /**
     * Agentic mode: sends the instruction, detects tool calls in the reply,
     * executes them, and re-prompts until the model produces a plain-text answer.
     */
    suspend fun agentAsync(instruction: String): String

    // ── Context ──────────────────────────────────────────────────────────────

    /** Returns the most recent [CompanionContext] snapshot. */
    fun getContext(): CompanionContext

    /** Refreshes the context from backing stores. */
    suspend fun refreshContextAsync()

    // ── History ───────────────────────────────────────────────────────────────

    /** The in-session conversation history (this session only, not persisted). */
    val history: List<CompanionTurn>

    // ── Feedback ──────────────────────────────────────────────────────────────

    /** Signal satisfaction with the last reply. */
    suspend fun signalFeedbackAsync(positive: Boolean, note: String? = null)

    // ── Proactive ─────────────────────────────────────────────────────────────

    /**
     * Raised when the Companion initiates contact without being prompted.
     */
    val proactiveEvents: Flow<CompanionProactiveEvent>
}
